cmd/autoscaler/main.go

```diff
@@ -147,15 +147,30 @@ func main() {
 		logger.Fatalw("Failed to start informers", zap.Error(err))
 	}
 
-	cc, selfIP := componentConfigAndIP(ctx, logger)
-	ctx = leaderelection.WithStandardLeaderElectorBuilder(ctx, kubeClient, cc)
+	cc := componentConfigAndIP(ctx)
 
 	// accept is the func to call when this pod owns the Revision for this StatMessage.
 	accept := func(sm asmetrics.StatMessage) {
 		collector.Record(sm.Key, time.Unix(sm.Stat.Timestamp, 0), sm.Stat)
 		multiScaler.Poke(sm.Key, sm.Stat)
 	}
-	f := statforwarder.New(ctx, logger, kubeClient, selfIP, bucket.AutoscalerBucketSet(cc.Buckets), accept)
+
+	var f *statforwarder.Forwarder
+	if b, bs, err := leaderelection.NewStatefulSetBucketAndSet(int(cc.Buckets)); err == nil {
```

Comment:
so can we ensure only one of the two will be configured?

Reply:
This is precisely how `WithDynamicLeaderElector` works today.  We key off of the statefulset env vars to determine whether that should be the source of bucket ownership.  I looked at whether we could make that return the BucketSet, but it only knows it for the StatefulSet variant and it is impossible to know ahead of time for the standard builder (unless you do something like autoscaling is doing).